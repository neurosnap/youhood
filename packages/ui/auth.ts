import * as h from 'react-hyperscript';
import { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actions, selectors as authSelectors } from '@youhood/auth';
const { signIn, register } = actions;
const { getAuthError } = authSelectors;
import { selectors } from '@youhood/user';
const { getCurrentUserId } = selectors;
import { AuthPayload, AuthError } from '@youhood/auth/types';
import { UserId } from '@youhood/user/types';

import { State } from '@youhood/web-app/types';
import {
  DropdownMenuButton,
  ErrorText,
  Input,
  Buttons,
  SignInMenuEl,
  DropdownMenuContainer,
  SigninMsgBase,
} from './ui';

type OnClick = (props: AuthPayload) => void;
interface IAuth {
  error?: AuthError;
  children?: any;
  onClick: OnClick;
  currentUserId: UserId;
  buttonText: string;
}

interface IAuthState {
  email: string;
  password: string;
}

export class Auth extends Component<IAuth, IAuthState> {
  static defaultProps = {
    buttonText: '',
    onClick: () => {},
    currentUserId: '',
    error: '',
  };

  state = {
    email: '',
    password: '',
  };

  handleEmail = (event: React.FormEvent<HTMLInputElement>) => {
    this.setState({ email: event.currentTarget.value });
  };

  handlePassword = (event: React.FormEvent<HTMLInputElement>) => {
    this.setState({ password: event.currentTarget.value });
  };

  handleClick = () => {
    const { email, password } = this.state;
    const { onClick, currentUserId } = this.props;
    onClick({ email, password, currentUserId });
  };

  render() {
    const { email, password } = this.state;
    const { error, buttonText } = this.props;
    const shouldDisableButton = !email || !password;

    return h(SignInMenuEl, [
      h(Input, {
        className: 'signin-email',
        type: 'text',
        placeholder: 'email address',
        value: email,
        onChange: this.handleEmail,
      }),
      h(Input, {
        className: 'signin-pass',
        type: 'password',
        placeholder: 'password',
        value: password,
        onChange: this.handlePassword,
      }),
      error ? h(ErrorText, error) : h(SigninMsgBase),
      h(Buttons, [
        h(
          DropdownMenuButton,
          {
            className: 'signin-btn',
            onClick: this.handleClick,
            disabled: shouldDisableButton,
          },
          buttonText,
        ),
      ]),
    ]);
  }
}

interface IAuthMenu {
  error: AuthError;
  onClick: OnClick;
  currentUserId: UserId;
}

export const createAuthMenu = (buttonText: string) => ({
  onClick,
  currentUserId,
  error,
}: IAuthMenu) => h(Auth, { onClick, currentUserId, error, buttonText });

export const SignInMenu = createAuthMenu('Sign In');
export const RegisterMenu = createAuthMenu('Register');

const mapStateToProps = (state: State) => ({
  currentUserId: getCurrentUserId(state),
  error: getAuthError(state),
});

const createConnectForAuth = (component: any, action: Function) =>
  connect(
    mapStateToProps,
    (dispatch: Function) => ({
      onClick: (payload: AuthPayload) => dispatch(action(payload)),
    }),
  )(component as any);

export const SignInMenuConn = createConnectForAuth(SignInMenu, signIn);
export const RegisterMenuConn = createConnectForAuth(RegisterMenu, register);

type Tab = 'Sign In' | 'Register';
interface IAuthMenuState {
  tabSelected: Tab;
}

const TabSelected = styled.div`
  text-decoration: underline;
`;
const TabInactive = styled.div`
  cursor: pointer;
  color: #ccc;

  :hover {
    color: #fff;
  }
`;
const TabMenu = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  height: 36px;
`;

export default class AuthMenu extends Component<null, IAuthMenuState> {
  state = {
    tabSelected: 'Sign In' as Tab,
  };

  setTab = (tab: Tab) => {
    const { tabSelected } = this.state;
    if (tab === tabSelected) return;
    this.setState({ tabSelected: tab });
  };

  render() {
    const { tabSelected } = this.state;
    const tabNames = ['Sign In', 'Register'];
    const tabs = tabNames.map((tab: Tab) => {
      const onClick = () => this.setTab(tab);
      const TabView = tab === tabSelected ? TabSelected : TabInactive;
      return h(TabView, { onClick, key: tab }, tab);
    });
    const Menu = tabSelected === 'Sign In' ? SignInMenuConn : RegisterMenuConn;

    return h(DropdownMenuContainer, [h(TabMenu, tabs), h(Menu)]);
  }
}
