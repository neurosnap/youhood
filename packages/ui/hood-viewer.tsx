import * as React from 'react';
import { connect } from 'react-redux';

import { HoodProps, HoodId, EditHoodPayload } from '@youhood/hood/types';
import { actions, selectors } from '@youhood/hood';
const { editHood } = actions;
const { getHoodSelected } = selectors;
import { selectors as userSelectors } from '@youhood/user';
const { getUserById, getCurrentUserId } = userSelectors;
import {
  selectors as voteSelectors,
  actions as voteActions,
} from '@youhood/vote';
import { VoteTypes } from '@youhood/vote/types';
const {
  getVoteCountByHood,
  didUserVoteOnHood,
  getUserVoteTypeForHood,
} = voteSelectors;
const { upvote, downvote, unvote } = voteActions;
import { selectors as tokenSelectors } from '@youhood/token';
const { getIsUserLoggedIn } = tokenSelectors;
import { User, WebState } from '@youhood/types';

import {
  HoodContainer,
  HeaderSmall,
  Votes,
  VoteUp,
  Voted,
  Actions,
  OverlayHeader,
  OverlayContainer,
  Link,
  Header,
  TextSmall,
} from './ui';
import { formatDate } from './date';
import HoodReport from './hood-report';

type UserId = string;

interface Props {
  canEdit: boolean;
  canUserVote: boolean;
  currentUserId: UserId;
  edit: (p: { hoodId: HoodId; edit: boolean }) => void;
  handleUnvote: (
    hoodId: HoodId,
    currentUserId: string,
    vote: VoteTypes,
  ) => void;
  handleUpvote: (hoodId: HoodId, currentUserId: string) => void;
  handleDownvote: (hoodId: HoodId, currentUserId: string) => void;
  hood: HoodProps;
  hoodId: HoodId;
  user: User;
  userVoted: boolean;
  votes: number;
  userVoteType: VoteTypes;
}

export class HoodViewer extends React.Component<Props> {
  handleEdit = (e: React.MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault();
    const { edit, hoodId } = this.props;
    edit({ hoodId, edit: true });
  };

  render() {
    const {
      hoodId,
      hood,
      user,
      canEdit,
      votes,
      userVoted,
      userVoteType,
      canUserVote,
      handleUpvote,
      handleDownvote,
      handleUnvote,
      currentUserId,
    } = this.props;
    if (!hoodId) {
      return null;
    }

    const actions = [];
    if (canEdit) {
      actions.push(
        <Link href="/edit" onClick={this.handleEdit}>
          Edit
        </Link>,
      );
    }

    const userUpvoted = userVoted && userVoteType === 'upvote';
    const userDownvoted = userVoted && userVoteType === 'downvote';
    const VoteUpState = userUpvoted ? Voted : VoteUp;
    const VoteDownState = userDownvoted ? Voted : VoteUp;
    const UserUpVoting = canUserVote ? (
      <VoteUpState
        onClick={() => {
          if (userUpvoted) {
            handleUnvote(hoodId, currentUserId, 'upvote');
            return;
          }

          handleUpvote(hoodId, currentUserId);
        }}
      >
        /\\
      </VoteUpState>
    ) : null;
    const UserDownVoting = canUserVote ? (
      <VoteDownState
        onClick={() => {
          if (userDownvoted) {
            handleUnvote(hoodId, currentUserId, 'downvote');
            return;
          }

          handleDownvote(hoodId, currentUserId);
        }}
      >
        \\/
      </VoteDownState>
    ) : null;

    const username = user.isTmp ? 'Anonymous' : user.email;
    return (
      <OverlayContainer>
        <OverlayHeader>
          <Header>Hood Viewer</Header>
          <Actions>{...actions}</Actions>
        </OverlayHeader>
        <HoodContainer>
          <Votes>
            {UserUpVoting}
            <Header>{votes}</Header>
            {UserDownVoting}
          </Votes>
          <div style={{ width: '85%' }}>
            <HeaderSmall>{hood.name}</HeaderSmall>
            <TextSmall>
              <div>User</div>
              <div>{username}</div>
            </TextSmall>
            <TextSmall>
              <div>Last Updated</div>
              <div>{formatDate(hood.updatedAt)}</div>
            </TextSmall>
            <TextSmall>
              <div>Created</div>
              <div>{formatDate(hood.createdAt)}</div>
            </TextSmall>
            <HoodReport hood={hood} />
          </div>
        </HoodContainer>
      </OverlayContainer>
    );
  }
}

export default connect(
  (state: WebState) => {
    const hood = getHoodSelected(state);
    const hoodId = hood.id;
    const userId = hood.userId;
    const user = getUserById(state, { id: userId });
    const currentUserId = getCurrentUserId(state);
    const didUserCreateHood = user && user.id === currentUserId;
    const userIsAuthenticated = getIsUserLoggedIn(state);
    const canUserVote = !didUserCreateHood && userIsAuthenticated;

    return {
      hoodId,
      hood,
      user: user || {},
      canEdit: didUserCreateHood,
      votes: getVoteCountByHood(state, { hoodId }),
      userVoted: didUserVoteOnHood(state, { hoodId, userId: currentUserId }),
      userVoteType: getUserVoteTypeForHood(state, {
        hoodId,
        userId: currentUserId,
      }),
      canUserVote,
      currentUserId,
    };
  },
  (dispatch: Function) => ({
    edit: (opts: EditHoodPayload) => dispatch(editHood(opts)),
    handleUpvote: (hoodId: HoodId, userId: UserId) =>
      dispatch(upvote({ hoodId, userId })),
    handleDownvote: (hoodId: HoodId, userId: UserId) =>
      dispatch(downvote({ hoodId, userId })),
    handleUnvote: (hoodId: HoodId, userId: UserId, voteType: VoteTypes) =>
      dispatch(unvote({ hoodId, userId, voteType })),
  }),
)(HoodViewer);
