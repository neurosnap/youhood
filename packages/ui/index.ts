import Overlay from './overlay';
import Navbar from './navbar';
import HoodBar from './hoodbar';

const smartComponents = {
  Overlay,
  Navbar,
  HoodBar,
};

export { smartComponents };
