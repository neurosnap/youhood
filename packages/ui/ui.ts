import styled from 'styled-components';

const successColor = '#28a745';
const errorColor = '#dc3545';
const primaryFontColor = '#fff';

// core

export const Header = styled.h1`
  font-size: 1.3rem;
`;

export const HeaderSmall = styled.h2`
  font-size: 1.2rem;
  margin-bottom: 0.5rem;
`;

export const Link = styled.a`
  margin: 0 0.2rem;
  color: ${primaryFontColor};
`;

export const LinkDanger = Link.extend`
  color: ${errorColor};
`;

export const LinkSuccess = Link.extend`
  color: ${successColor};
`;

export const InputBase = styled.input`
  margin: 0;
  outline: none;
`;

export const TextSmall = styled.div`
  font-size: 0.8rem;
  display: flex;
  justify-content: space-between;
`;

// derived

export const InputOverlay = InputBase.extend`
  padding: 0.375rem 0.75rem;
  color: #495057;
  background-color: #fff;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  width: 100%;
`;

const OverlaySection = styled.div`
  background-color: #fff;
  border-radius: 0.3rem;
  padding: 0.6rem;
`;

export const HoodContainer = OverlaySection.extend`
  display: flex;
`;
export const HoodEditorContainer = OverlaySection;

export const HoodSelectionContainer = OverlaySection.extend`
  display: flex;
  flex-direction: column;
  margin-bottom: 0.5rem;
`;

export const HoodSelectionItem = styled.div`
  display: flex;
  align-items: center;
`;

export const HoodVisibility = styled.i`
  cursor: pointer;
`;

export const Votes = styled.div`
  padding: 0 0.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

export const VoteUp = styled.i`
  cursor: pointer;
`;

export const Voted = VoteUp.extend`
  color: orange;
`;

export const Actions = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const EditorActions = Actions.extend`
  margin-top: 0.5rem;
`;

export const HoodBaseItem = styled.div`
  cursor: pointer;
  padding: 0.3rem 0.4rem;
  border-radius: 0.3rem;
  flex: 1;
  margin-left: 0.3rem;
`;

export const HoodListItem = HoodBaseItem.extend`
  :hover {
    background-color: rgba(46, 139, 87, 0.5);
  }
`;

export const HoodListItemSelected = HoodBaseItem.extend`
  background-color: rgb(255, 250, 205);
`;

export const OverlayHeader = styled.div`
  font-size: 1rem;
  color: ${primaryFontColor};
  display: flex;
  align-items: center;
  margin-bottom: 0.6rem;
`;

export const OverlayContainer = styled.div`
  width: 90%;
  margin-top: 0.5rem;
`;

export const NavHover = styled.button`
  cursor: pointer;
  height: 100%;
  display: flex;
  align-items: center;
  padding: 0 0.4rem;
  margin: 0 0.4rem;
  background: inherit;
  color: inherit;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: inherit;

  :hover:enabled {
    background-color: rgba(255, 255, 255, 0.25);
  }

  :disabled {
    color: #ccc;
  }
`;

export const HoodBarButton = NavHover.extend`
  font-size: 0.8rem;
`;

export const Input = InputBase.extend`
  margin: 0.4rem;
  padding: 0.4rem;
  border-radius: 0.1rem;
  height: 3rem;
  border: none;
  background-image: none;
  background-color: rgba(255, 255, 255, 0.15);
  box-shadow: none;
  width: 90%;
  color: ${primaryFontColor};
  font-size: 1rem;

  :hover {
    background-color: rgba(255, 255, 255, 0.25);
  }

  ::-webkit-input-placeholder {
    color: #eee;
  }
`;

export const SigninMsgBase = styled.div`
  height: 2rem;
`;

export const ErrorText = SigninMsgBase.extend`
  color: #721c24;
  background-color: #f8d7da;
  font-size: 0.8rem;
  padding: 0 0.5rem;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const Buttons = styled.div`
  margin-top: 0.4rem;
  display: flex;
`;

export const Menu = NavHover.extend`
  height: 2rem;
`;

export const Nav = styled.div`
  display: flex;
  align-items: center;
  height: 56px;
  background: #4285f4;
  box-sizing: border-box;
  top: 0;
  width: 100vw;
  color: ${primaryFontColor};
`;

export const NavContent = styled.div`
  width: 100%;
  margin: 0 1.5rem;
  display: flex;
  align-items: center;
`;

export const NavLeft = styled.div`
  display: flex;
  align-items: center;
`;

export const NavRight = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2.2rem;
  width: 75%;
`;

export const Search = styled.div`
  width: 40%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.15);
  border-radius: 0.1rem;
  display: flex;
  align-items: center;

  :hover {
    background-color: rgba(255, 255, 255, 0.25);
  }
`;

export const SearchInput = InputBase.extend`
  font-size: 1rem;
  background: none;
  border: none;
  box-sizing: border-box;
  height: 100%;
  flex: 1;
  color: ${primaryFontColor};

  ::-webkit-input-placeholder {
    color: ${primaryFontColor};
  }
`;

export const SearchIcon = styled.i`
  margin: 0 0.5rem;
`;

export const Brand = styled.div`
  font-size: 1.1rem;
  margin: 0 1.5rem;
`;

export const SignInMenuEl = styled.div`
  height: 80%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const DropdownMenuContainer = styled.div`
  position: absolute;
  background: #4285f4;
  top: 46px;
  right: 9px;
  z-index: 401;
  width: 310px;
  height: 260px;
`;

export const DropdownMenuButton = NavHover.extend`
  height: 2rem;
`;

export const Points = styled.div`
  color: ${primaryFontColor};
  background-color: limegreen;
  border-radius: 0.3rem;
  padding: 0.1rem 0.3rem;
  font-size: 0.8rem;
`;

export const SignInContainer = styled.div`
  margin-left: 0;
  position: relative;
  height: 100%;
`;

export const SignInEl = styled.div`
  text-decoration: none;
  color: ${primaryFontColor};

  a {
    text-decoration: none;
    color: ${primaryFontColor};
  }
`;
