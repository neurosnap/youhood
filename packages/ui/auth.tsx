import * as React from 'react';
import styled from 'styled-components';

import theme from '@youhood/theme';

import { DropdownMenuContainer } from './ui';
import SignIn from './signin';
import Register from './register';

type Tab = 'Sign In' | 'Register';
interface IAuthMenuState {
  tabSelected: Tab;
}

const TabSelected = styled.div`
  color: ${theme.palette.font.primary};
  text-decoration: underline;
`;
const TabInactive = styled.div`
  cursor: pointer;
  color: ${theme.palette.font.disabled};

  :hover {
    color: ${theme.palette.font.primary};
  }
`;
const TabMenu = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  height: 36px;
`;

export class AuthMenu extends React.Component<any, IAuthMenuState> {
  state = {
    tabSelected: 'Sign In' as Tab,
  };

  setTab = (tab: Tab) => {
    const { tabSelected } = this.state;
    if (tab === tabSelected) return;
    this.setState({ tabSelected: tab });
  };

  render() {
    const { tabSelected } = this.state;
    const tabNames: Tab[] = ['Sign In', 'Register'];
    const tabs = tabNames.map((tab: Tab) => {
      const onClick = () => this.setTab(tab);
      const TabView = tab === tabSelected ? TabSelected : TabInactive;
      return (
        <TabView onClick={onClick} key={tab}>
          {tab}
        </TabView>
      );
    });
    const Menu = tabSelected === 'Sign In' ? SignIn : Register;

    return (
      <div>
        <TabMenu>{tabs}</TabMenu>
        <Menu />
      </div>
    );
  }
}

export default () => (
  <DropdownMenuContainer>
    <AuthMenu />
  </DropdownMenuContainer>
);
