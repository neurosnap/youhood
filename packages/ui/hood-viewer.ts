import { Component } from 'react';
import { connect } from 'react-redux';
import * as h from 'react-hyperscript';

import { HoodProps, HoodId, EditHoodPayload } from '@youhood/hood/types';
import { User, UserId } from '@youhood/user/types';
import { actions, selectors, utils } from '@youhood/hood';
const { editHood } = actions;
const { createHood } = utils;
const { getHoodSelected } = selectors;
import { selectors as userSelectors } from '@youhood/user';
const { getUserById, getCurrentUserId } = userSelectors;
import {
  selectors as voteSelectors,
  actions as voteActions,
} from '@youhood/vote';
const { getVoteCountByHood, didUserVoteOnHood } = voteSelectors;
const { vote, unvote } = voteActions;
import { selectors as authSelectors } from '@youhood/auth';
const { isUserAuthenticated } = authSelectors;
import { State } from '@youhood/web-app/types';

import {
  HoodContainer,
  HeaderSmall,
  Votes,
  VoteUp,
  Voted,
  Actions,
  OverlayHeader,
  OverlayContainer,
  Link,
  Header,
  TextSmall,
} from './ui';
import { formatDate } from './date';

const noop = () => {};

interface Props {
  canEdit?: boolean;
  canUserVote?: boolean;
  currentUserId?: UserId;
  edit?: Function;
  handleUnvote?: Function;
  handleVote?: Function;
  hood: HoodProps;
  hoodId: HoodId;
  user?: User;
  userVoted?: boolean;
  votes?: number;
}

interface DefaultProps {
  canEdit: boolean;
  canUserVote: boolean;
  currentUserId: UserId;
  edit: Function;
  handleUnvote: Function;
  handleVote: Function;
  hood: HoodProps;
  user: User;
  userVoted: boolean;
  votes: number;
}

export class HoodViewer extends Component<Props> {
  static defaultProps: DefaultProps = {
    canEdit: false,
    canUserVote: false,
    currentUserId: '',
    edit: noop,
    handleUnvote: noop,
    handleVote: noop,
    hood: createHood({ id: '123' }),
    user: {
      email: 'Unknown',
      id: '',
      createdAt: '',
      isTmp: true,
    },
    userVoted: false,
    votes: 0,
  };

  handleEdit = (e: React.MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault();
    const { edit, hoodId } = this.props;
    edit({ hoodId, edit: true });
  };

  render() {
    const {
      hoodId,
      hood,
      user,
      canEdit,
      votes,
      userVoted,
      canUserVote,
      handleVote,
      handleUnvote,
      currentUserId,
    } = this.props;
    if (!hoodId) {
      return null;
    }

    const actions = [];
    if (canEdit) {
      actions.push(
        h(Link, { href: '/edit', onClick: this.handleEdit }, 'Edit'),
      );
    }

    const VoteState = userVoted ? Voted : VoteUp;
    const UserVoting = canUserVote
      ? h(VoteState, {
          className: 'fa fa-angle-up',
          onClick: () => {
            if (userVoted) {
              handleUnvote(hoodId, currentUserId);
              return;
            }

            handleVote(hoodId, currentUserId);
          },
        })
      : null;

    const username = user.isTmp ? 'Anonymous' : user.email;
    return h(OverlayContainer, [
      h(OverlayHeader, [h(Header, 'Hood Viewer'), h(Actions, actions)]),
      h(HoodContainer, [
        h(Votes, [UserVoting, h(Header, votes)]),
        h('div', { style: { width: '85%' } }, [
          h(HeaderSmall, hood.name),
          h(TextSmall, [h('div', 'User'), h('div', username)]),
          h(TextSmall, [
            h('div', 'Last Updated'),
            h('div', formatDate(hood.updatedAt)),
          ]),
          h(TextSmall, [
            h('div', 'Created'),
            h('div', formatDate(hood.createdAt)),
          ]),
        ]),
      ]),
    ]);
  }
}

export default connect(
  (state: State) => {
    const hood = getHoodSelected(state);
    const hoodId = hood.id;
    const userId = hood.userId;
    const user = getUserById(state, { id: userId });
    const currentUserId = getCurrentUserId(state);
    const didUserCreateHood = user && user.id === currentUserId;
    const userIsAuthenticated = isUserAuthenticated(state);
    const canUserVote = !didUserCreateHood && userIsAuthenticated;

    return {
      hoodId,
      hood,
      user: user || {},
      canEdit: didUserCreateHood,
      votes: getVoteCountByHood(state, { hoodId }),
      userVoted: didUserVoteOnHood(state, { hoodId, userId }),
      canUserVote,
      currentUserId,
    };
  },
  (dispatch: Function) => ({
    edit: (opts: EditHoodPayload) => dispatch(editHood(opts)),
    handleVote: (hoodId: HoodId, userId: UserId) =>
      dispatch(vote({ hoodId, userId })),
    handleUnvote: (hoodId: HoodId, userId: UserId) =>
      dispatch(unvote({ hoodId, userId })),
  }),
)(HoodViewer as any);
