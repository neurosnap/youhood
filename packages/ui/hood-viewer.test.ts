import { mount } from 'enzyme';
import * as h from 'react-hyperscript';

import { HoodViewer } from './hood-viewer';
import { Voted, VoteUp, Link } from './ui';

const tmpUser = {
  email: '123',
  id: '123',
  isTmp: true,
};
const regUser = {
  email: 'eric@cool.com',
  id: '123',
  isTmp: false,
};
const hood = {
  id: '321',
  city: 'Ann Arbor',
  state: 'Michigan',
  county: 'Washtenaw',
  createdAt: '2018-01-01',
  updatedAt: '2018-01-01',
  userId: tmpUser.id,
  name: 'Da Hood',
};

describe('HoodView', () => {
  describe('when the user is temporary', () => {
    it('should render `Anonymous` user name', () => {
      const tree = mount(
        h(HoodViewer, {
          hood,
          hoodId: hood.id,
          currentUserId: tmpUser.id,
          user: tmpUser,
        }),
      );

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when the user has registered', () => {
    it('should render the email as user', () => {
      const tree = mount(
        h(HoodViewer, {
          hood,
          hoodId: hood.id,
          user: regUser,
        }),
      );

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when the user can vote on hood', () => {
    describe('when the user has not voted yet', () => {
      const handleVote = jest.fn();
      const tree = mount(
        h(HoodViewer, {
          hood,
          hoodId: hood.id,
          user: regUser,
          userVoted: false,
          canUserVote: true,
          handleVote,
          currentUserId: regUser.id,
        }),
      );
      const vote = tree.find(VoteUp);

      it('should display the upvote arrow', () => {
        expect(vote.length).toEqual(1);
      });

      describe('when the user clicks the upvote arrow', () => {
        it('should call handleVote', () => {
          vote.simulate('click');
          expect(handleVote).toHaveBeenCalledWith('321', regUser.id);
        });
      });
    });

    describe('when the user has voted', () => {
      const handleUnvote = jest.fn();
      const tree = mount(
        h(HoodViewer, {
          hood,
          hoodId: hood.id,
          user: regUser,
          userVoted: true,
          canUserVote: true,
          handleUnvote,
          currentUserId: regUser.id,
        }),
      );
      const vote = tree.find(Voted);

      it('should display the voted arrow', () => {
        expect(vote.length).toEqual(1);
      });

      describe('when the user clicks the upvote arrow', () => {
        it('should call handleUnvote', () => {
          vote.simulate('click');
          expect(handleUnvote).toHaveBeenCalledWith('321', regUser.id);
        });
      });
    });
  });

  describe('when the user can edit the hood', () => {
    describe('when the user is temporary', () => {
      it('should display edit button', () => {
        const tree = mount(
          h(HoodViewer, {
            hood,
            hoodId: hood.id,
            user: tmpUser,
            canEdit: true,
          }),
        );

        expect(tree.find(Link).length).toEqual(1);
      });
    });
  });
});
