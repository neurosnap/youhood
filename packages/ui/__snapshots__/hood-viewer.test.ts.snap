// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoodView when the user has registered should render the email as user 1`] = `
.c2 {
  font-size: 1.3rem;
}

.c6 {
  font-size: 1.2rem;
  margin-bottom: 0.5rem;
}

.c7 {
  font-size: 0.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  background-color: #fff;
  border-radius: 0.3rem;
  padding: 0.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  padding: 0 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 1rem;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 0.6rem;
}

.c0 {
  width: 90%;
  margin-top: 0.5rem;
}

<HoodViewer
  canEdit={false}
  canUserVote={false}
  currentUserId=""
  edit={[Function]}
  handleDownvote={[Function]}
  handleUnvote={[Function]}
  handleUpvote={[Function]}
  hood={
    Object {
      "city": "Ann Arbor",
      "county": "Washtenaw",
      "createdAt": "2018-01-01",
      "id": "321",
      "name": "Da Hood",
      "state": "Michigan",
      "updatedAt": "2018-01-01",
      "userId": "123",
    }
  }
  hoodId="321"
  user={
    Object {
      "email": "eric@cool.com",
      "id": "123",
      "isTmp": false,
    }
  }
  userVoteType={null}
  userVoted={false}
  votes={0}
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          <styled.h1>
            <h1
              className="c2"
            >
              Hood Viewer
            </h1>
          </styled.h1>
          <styled.div>
            <div
              className="c3"
            />
          </styled.div>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c4"
        >
          <styled.div>
            <div
              className="c5"
            >
              <styled.h1>
                <h1
                  className="c2"
                >
                  0
                </h1>
              </styled.h1>
            </div>
          </styled.div>
          <div
            style={
              Object {
                "width": "85%",
              }
            }
          >
            <styled.h2>
              <h2
                className="c6"
              >
                Da Hood
              </h2>
            </styled.h2>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  User
                </div>
                <div>
                  eric@cool.com
                </div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  Last Updated
                </div>
                <div>
                  01-01-2018 00:00 am
                </div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  Created
                </div>
                <div>
                  01-01-2018 00:00 am
                </div>
              </div>
            </styled.div>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</HoodViewer>
`;

exports[`HoodView when the user is temporary should render \`Anonymous\` user name 1`] = `
.c2 {
  font-size: 1.3rem;
}

.c6 {
  font-size: 1.2rem;
  margin-bottom: 0.5rem;
}

.c7 {
  font-size: 0.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  background-color: #fff;
  border-radius: 0.3rem;
  padding: 0.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  padding: 0 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 1rem;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 0.6rem;
}

.c0 {
  width: 90%;
  margin-top: 0.5rem;
}

<HoodViewer
  canEdit={false}
  canUserVote={false}
  currentUserId="123"
  edit={[Function]}
  handleDownvote={[Function]}
  handleUnvote={[Function]}
  handleUpvote={[Function]}
  hood={
    Object {
      "city": "Ann Arbor",
      "county": "Washtenaw",
      "createdAt": "2018-01-01",
      "id": "321",
      "name": "Da Hood",
      "state": "Michigan",
      "updatedAt": "2018-01-01",
      "userId": "123",
    }
  }
  hoodId="321"
  user={
    Object {
      "email": "123",
      "id": "123",
      "isTmp": true,
    }
  }
  userVoteType={null}
  userVoted={false}
  votes={0}
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          <styled.h1>
            <h1
              className="c2"
            >
              Hood Viewer
            </h1>
          </styled.h1>
          <styled.div>
            <div
              className="c3"
            />
          </styled.div>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c4"
        >
          <styled.div>
            <div
              className="c5"
            >
              <styled.h1>
                <h1
                  className="c2"
                >
                  0
                </h1>
              </styled.h1>
            </div>
          </styled.div>
          <div
            style={
              Object {
                "width": "85%",
              }
            }
          >
            <styled.h2>
              <h2
                className="c6"
              >
                Da Hood
              </h2>
            </styled.h2>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  User
                </div>
                <div>
                  Anonymous
                </div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  Last Updated
                </div>
                <div>
                  01-01-2018 00:00 am
                </div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c7"
              >
                <div>
                  Created
                </div>
                <div>
                  01-01-2018 00:00 am
                </div>
              </div>
            </styled.div>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</HoodViewer>
`;
